File: 735-asteroid-collision/asteroid-collision.cpp
class Solution {
public:
    vector<int> asteroidCollision(vector<int>& asteroids) {
        stack<int>st;
        for(int &i:asteroids){
            while(!st.empty() && i<0 && st.top()>0){
                int sum=i+st.top();
                if(sum<0) st.pop();
                else if(sum>0) i=0;
                else{
                    st.pop();
                    i=0;
                }
            }
            if(i!=0) st.push(i);
        }
        int s=st.size();
        vector<int>ans(s);
        int i=s-1;
        while(!st.empty()){
            ans[i]=st.top();
            st.pop();
            i--;
        }
        return ans;
    }
};

File: 735-asteroid-collision/asteroid-collision.py
class Solution(object):
    def asteroidCollision(self, asteroids):
        """
        :type asteroids: List[int]
        :rtype: List[int]
        """
        stack = []

        for asteroid in asteroids:
            # Handle collisions
            while stack and asteroid < 0 < stack[-1]:
                # Case 1: right asteroid is smaller -> pop it
                if stack[-1] < -asteroid:
                    stack.pop()
                    continue
                # Case 2: both are same size -> both explode
                elif stack[-1] == -asteroid:
                    stack.pop()
                # Case 3: right asteroid is bigger -> left asteroid explodes
                break
            else:
                # No collision or all resolved -> push this asteroid
                stack.append(asteroid)

        return stack
